book-content/chapters/11-annotations-and-assertions.md

```diff
@@ -0,0 +1,1196 @@
+
+1. Ensure that the value passed to the variable matches the annotation.
+2. Forget about the value's type.
+
+This has some benefits - we can add new keys to `config` and TypeScript won't complain:
+
+```typescript
+config.primary = "red";
+```
+
+But this isn't really what we want - this is a config object that shouldn't be changed.
+
+### With No Annotation, The Value Wins
+
+One way to get around this would be to drop the variable annotation.
+
+```typescript
+const config = {
+  foreground: { r: 255, g: 255, b: 255 },
+  background: { r: 0, g: 0, b: 0 },
+  border: "transparent",
+};
+```
+
+Because there's no variable annotation, `config` is inferred as the type of the value provided.
+
+But now we've lost the ability to check that the `Color` type is correct. We can add a `number` to the `foreground` key and TypeScript won't complain:
+
+```typescript
+const config = {
+  foreground: 123,
+};
+```
+
+So it seems we're at an impasse. We both want to infer the type of the value, but also constrain it to be a certain shape.
+
+### Annotating Values With `satisfies`
+
+The `satisfies` operator is a way to tell TypeScript that a value must satisfy certain criteria, but still allow TypeScript to infer the type.
+
+Let's use it to make sure our `config` object has the right shape:
+
+```typescript
+const config = {
+  foreground: { r: 255, g: 255, b: 255 },
+  background: { r: 0, g: 0, b: 0 },
+  border: "transparent",
+} satisfies Record<string, Color>;
+```
+
+Now, we get the best of both worlds. This means we can access the keys without any issues:
+
+```typescript
+config.foreground.r;
+
+config.border.toUpperCase();
+```
+
+But we've also told TypeScript that `config` must be a `Record` with a string key and a `Color` value. If we try to add a key that doesn't match this shape, TypeScript will show an error:
+
+```ts twoslash
+// @errors: 2322
+type Color =
+  | string
+  | {
+      r: number;
+      g: number;
+      b: number;
+    };
+
+// ---cut---
+const config = {
+  primary: 123,
+} satisfies Record<string, Color>;
+```
+
+Of course, we have now lost the ability to add new keys to `config` without TypeScript complaining:
+
+```ts twoslash
+// @errors: 2339
+type Color =
+  | string
+  | {
+      r: number;
+      g: number;
+      b: number;
+    };
+
+const config = {} satisfies Record<string, Color>;
+// ---cut---
+config.somethingNew = "red";
+```
+
+Because TypeScript is now inferring `config` as _just_ an object with a fixed set of keys.
+
+Let's recap:
+
+- When you use a variable annotation, the variable's type wins.
+- When you don't use a variable annotation, the value's type wins.
+- When you use `satisfies`, you can tell TypeScript that a value must satisfy certain criteria, but still allow TypeScript to infer the type.
```

Comment:
which has some notable surprising caveats at times, like:
```ts
type User = { name: string }

const arr = [] satisfies User[]
//    ^? const arr: never[]
```

The `satisfies` operator is often described as "providing contextual information without changing the type of the expression" (which tbh... isn't probably that helpful to newcomers)